sdk/overview.md

```diff
@@ -0,0 +1,127 @@
+Allow flexibility on network layer etc.
+How to prevent users from including large binary when they need single service.
+ 
+5. Adding common services and addons for all services
+Functionalities like network handling, logging, monitoring and audit, error handling, analytics.
+
+6. Developer tooling 
+How to build or provide (plugins, extensions) to service
+
+7. Size constraints 
+For example method limit on android
+
+8. Platform language
+Supporting multiple flavours (languages per platform) - Swift vs ObjectiveC. Java vs Kotlin etc.
+
+## Proposed Solution
+
+Modular SDK For SDK that can contain multiple non related services 
+
+### Overview
+
+Modular architecture is a software design technique that emphasizes separating the functionality of a program into independent, interchangeable modules, such that each contains everything necessary to execute only one aspect of the desired functionality.
+
+Modular architecture is well known as practice for the server side development, but it can be also applied when building mobile SDK. In modular mobile SDK, code can be divided into separate packages that can be used separately by specific platform package manager. Each package depends on core package that needs to be added as separate dependency.
+
+![Diagram](./Diagram.png)
+
+This technique applies especially for large SDK’s that will implement clients for multiple not strictly correlated services. 
+
+Having top level SDK allows developers to provide fluent api (different forms of the top level api that can be swapped and extended). Modular architecture also enforces good programming patterns. Networking, Logging etc. are loosely coupled with the service which allows developers to swap implementations with their own favorite network or logging library.
+
+#### Advantages of modular SDK
+
+- Easy to maintain as common elements can be extracted to base layer
+- Easy to scale while providing uniform way of integration with services
+- Simplicity when maintaining external dependencies
+- Possibility to have multiple variants/extensions to the services without including them into source code.
+- Modularity helps to separate common elements with all services
+- More flexible (SDK can be distributed both as bundle or as single services)
+- More centralized approach for services 
+- Simplicity of releases - out of the box when working with tools like lerna etc.
+- Simplicity of analytics and other cross service components.
+- Simplified documentation (assuming that packages are in mono repository)
+
+####  Possible side effects
+
+- Grouping services together may make it harder to contribute to the individual services.
+- Modular SDK’s are easy to maintain in platform specific mono repositories, which may not be acceptable in some cases.
+
+### General requirements 
+
+SDK should relay on reactive (RX*) api used for internal communication between SDK components. On each platform can be implemented using RXJS, RXJava etc.
```

Comment:
/s/relay/rely

So one core question to this is do we want to expose Rx* APIs in our APIs?  In general I am a "no" on that question.  Rx can be a implementation detail however.

In Android a lot of the community uses RxJava and RxAndroid.  If we include it as a dependency then we end up having to be very careful about dependency management.  There are two ways that come to mind to handle this.  

The first way is that we enforce some kind of default dependency version (Maven BOM) to avoid potential collisions.  It is overrideable so savvy developers can include their own libraries instead of ours and test for breakages themselves.  I think this should be the default. 

The second way is we shade the Rx* namespaces.  This way we don't have to expose ANY information about RxJava and users' are free to choose whichever version they want.  This is a lot more complex however.

Can people chime in about other platforms? @danielpassos do you have thoughts on iOS?

Reply:
> So one core question to this is do we want to expose Rx* APIs in our APIs?

No. It will be internal. 
This is really valuable comment. Maybe we should keep SDK's very small and avoid using user level libraries like rx? Technically we can achieve the same results. 

The last thing we want is to have some poor user experience and dependency collisions on users projects. This will be valid for both Android and IOS. 

For IOS: http://blog.sigmapoint.pl/avoiding-dependency-collisions-in-ios-static-library-managed-by-cocoapods/

``Write in docs which dependencies are required - you can explicitly say that you require e.g. AFNetworking in version 1.3.3 and basically force user of your library to provide such version during build time``

Reply:
I'm thinking to remove this restriction and do not mention RX for the moment, but the same problem may happen again with other dependencies.

Reply:
+1 on removing this. It is the implementation details for each SDK, not a high level architecture design.